pkg/statemachine/stateMachine.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package statemachine
+
+type StateDirection int64
+
+const (
+	Forward StateDirection = iota
+	Backward
+	Stop
+)
+
+type StateMachine struct {
+	index    int
+	states   []string
+	finished bool
+}
+
+func NewStateMachine(states []string) *StateMachine {
+	return &StateMachine{
+		states: states,
+	}
+}
+
+func (sm *StateMachine) CurrentState() string {
+	if sm.index < 0 || sm.index >= len(sm.states) {
+		return ""
+	}
+	return sm.states[sm.index]
+}
+
+func (sm *StateMachine) NextState(direction StateDirection) string {
+	switch direction {
+	case Forward:
+		sm.index++
+	case Backward:
+		sm.index--
+	default:
+		return ""
+	}
+	if sm.index == len(sm.states) {
```

Comment:
This looks like a very important invariant: as long as I do not misunderstand, for this to always be true, the last element of `StateDirection` needs always to be `Stop`, so we might want to document this clearly somewhere.

Reply:
I think I don't understand the invariant you refer to. I am not seeing a proposition involving StateDirection here.

Reply:
Your statemachine is implicitly assuming that the last index is always a stop. This is not necessarily the way a state machine works, where each state has its own label and the current state isn't identified by an index but by the actual label, to avoid having to interpret the index (e.g. `state == STATE_FORWARD`, `state == STATE_BACKWARD`, `state == INIT`, `state == PAUSE`, `state == EXIT`  etc.)

Reply:
yes, last state is assumed as ending state. it is now documented in the comment for how the machine works in the module.
and, I avoided giving meaning to any state because wanted the user to has his own semantics and not depend on his states. I only have the end state assumption, the possibility to stop the machine, and the backward/forward state transitions.